### DISCUSSION TITLE
Node API documentation?

### DISCUSSION THREAD

**Discussion by slorber** [browserslist/browserslist#846] [General]
Hi,

I tried to look around and wasn't sure to know where the Node API is documented. 

Also, if it's not documented, can we expect that you maintain semver retrocompatibility on it?

--- 

## Context

For the Docusaurus framework, I'm working on adding opt-in experimental Rspack support. 

It has built-in LightningCSS support with a slightly different API.

- Official JS binding: `targets` is something like `browserslistToTargets(browserslist(queries))`
- Build-in: `targets` is a list of browserslist queries directly (they did that to avoid installing an extra conversion package exposing `browserslistToTargets`)

It seems I'm able to read the current browserslist queries with the following code:

```ts
function getBrowserslistQueries(): string[] {
  const configFile = browserslist.findConfigFile(process.cwd());
  const queries = browserslist.loadConfig({path: configFile}) ?? [...browserslist.defaults];
  return queries;
}
```

I was wondering if this code is safe for framework usage with unpinned dependency, and if there's not a simpler way to achieve the same?

**Comment by ai** [comment]
Here is Node.js API docs: https://github.com/browserslist/browserslist?tab=readme-ov-file#js-api

You should use something like:

```js
function getBrowserslistQueries(): string[] {
  return browserslist();
}
```

Calling `findConfigFile` and `loadConfig` is very low-level methods, which you should avoid.

**Reply by slorber** [reply]
Thanks

The problem is that it's not exactly the same.

What I want to retrieve is this:

```js
[
      ">0.5%",
      "not dead",
      "not op_mini all"
]
```

Not the list of browsers returned by `browserslist()`

Rspack has built-in LightningCSS / browserslist support (I guess through [browserslist-rs](https://github.com/browserslist/browserslist-rs) ?) and accepts queries directly:

https://rspack.dev/guide/features/builtin-lightningcss-loader#targets

And it may not be in sync with the JS browserslist version I'm using. Passing the return of `browserslist()` to it gives me the following error for example:

>   [cause]: Error: Failed to parse browserslist: unknown version '128' of browser 'edge'

**Reply by slorber** [reply]
@chenjiahan said to me:

> Rspack only accepts browserslist query as the target and does not support the original LightningCSS option, this is by design.
> The original LightningCSS option relies on the lightningcss.browserslistToTargets method, which requires the user to install a lightningcss package, and we do not want to force Rspack users to install the package.

**Reply by ai** [reply]
Hm, got it. 

Yes, we don't have specific API for that case. Maybe we can add it since I think it can be useful for other cases.

**Reply by chenjiahan** [reply]
I made a minimal package to load browserslist config, it's used by Rsbuild. See: https://github.com/rspack-contrib/browserslist-load-config

**Reply by ai** [reply]
Nice! I will add method but also add a link to this helper

**Reply by slorber** [reply]
Thanks

Considering we already have browserslist installed I'd prefer to use the "official" package instead of adding an extra one.

> I will add method

So you plan to introduce another method as a public API surface and don't plan to simply document the existing ones?

Looks to me like just exposing `loadConfig` with `{path: process.cwd()}` by default could do the job.

**Reply by ai** [reply]
> So you plan to introduce another method as a public API surface and don't plan to simply document the existing ones?

Yes, it will be better to keep algorithm in sync with rest of Browserslist.

Reimplantating it could lead to a different behaviours. But I will add it to docs if course.

**Comment by ai** [comment]
The simpler way to do it is just to use `loadConfig()` it could find config automatically:

```js
browserslist.loadConfig({path: process.cwd()}) ?? [...browserslist.defaults]
```

I improved docs 9320799

**Reply by slorber** [reply]
Thanks, yes I figured this out yesterday, didn't initially understand what path was

https://github.com/facebook/docusaurus/pull/10402/commits/8093a2185f3517b16525087d1008269d7b5b95c2

BTW maybe it could use `process.cwd()` by default if not provided? 

`browserslist.loadConfig()` would be a quite convenient public API.
